### DISCUSSION TITLE
Multivalued attributes not being fetched

### DISCUSSION THREAD

**Discussion by javac9** [openmdx/openmdx-example#15] [General]
Hello, I've already posted a question in OpenCRX's Discussions, but since the example projects are also affected, posting it also here with example outputs.

Post in OpenCrx discussions: https://github.com/opencrx/opencrx/discussions/123
(the issue described there is actually not limited to getting single `CodeValueEntry`s but it's rather present when fetching any single entity object.)

---------------------------------------------------------

Now on the topic of OpenMDX example projects (in this case: Workshop example)
The behaviour is the same when using Swagger, but using XML directly in browser is somewhat simpler for demonstration.

A `Project` is added to the `Segment` so: `http://localhost:8080/openmdx-workshop-rest/org.openmdx.example.workshop1/provider/Workshop/segment/Standard/project` will return a `ResultSet` object with corresponding multivalued attributes (those contained in "PRFX_PROJECT_" table) of each `Project`, such as `createdBy` and `modifiedBy`:
```
<org.openmdx.kernel.ResultSet href="http://localhost:8080/openmdx-workshop-rest/org.openmdx.example.workshop1/provider/Workshop/segment/Standard/project" hasMore="false" total="1">
    <org.openmdx.example.workshop1.Project href="http://localhost:8080/openmdx-workshop-rest/org.openmdx.example.workshop1/provider/Workshop/segment/Standard/project/SICLE" version="bW9kaWZpZWRBdD0yMDIxLTA3LTI5VDA5OjI4OjU4LjMyOTAwMFo=">
        <createdBy>
            <_item index="0">
                <![CDATA[ junit ]]>
            </_item>
        </createdBy>
        <description>
            <![CDATA[ openMDX/openCRX Workshop (2006-02-27 to 2006-03-01 ]]>
        </description>
        <modifiedBy>
            <_item index="0">
                <![CDATA[ junit ]]>
            </_item>
        </modifiedBy>
        <identity>
            <![CDATA[ xri://@openmdx*org.openmdx.example.workshop1/provider/Workshop/segment/Standard/project/SICLE ]]>
        </identity>
        <modifiedAt>2021-07-29T09:28:58.329Z</modifiedAt>
        <name>
            <![CDATA[ Workshop ]]>
        </name>
        <createdAt>2021-07-29T09:28:58.329Z</createdAt>
    </org.openmdx.example.workshop1.Project>
</org.openmdx.kernel.ResultSet>
```

But... if we try: `http://localhost:8080/openmdx-workshop-rest/org.openmdx.example.workshop1/provider/Workshop/segment/Standard/project/SICLE` which is the only project in the Workshop example, what we get is:
```
<org.openmdx.example.workshop1.Project href="http://localhost:8080/openmdx-workshop-rest/org.openmdx.example.workshop1/provider/Workshop/segment/Standard/project/SICLE" version="bW9kaWZpZWRBdD0yMDIxLTA3LTI5VDA5OjI4OjU4LjMyOTAwMFo=">
    <description>
        <![CDATA[ openMDX/openCRX Workshop (2006-02-27 to 2006-03-01 ]]>
    </description>
    <identity>
        <![CDATA[ xri://@openmdx*org.openmdx.example.workshop1/provider/Workshop/segment/Standard/project/SICLE ]]>
    </identity>
    <modifiedAt>2021-07-29T09:28:58.329Z</modifiedAt>
    <name>
        <![CDATA[ Workshop ]]>
    </name>
    <createdAt>2021-07-29T09:28:58.329Z</createdAt>
</org.openmdx.example.workshop1.Project>
```
Note how the `createdBy` and `modifiedBy` attributes are missing; actually any other attributes persisted in `PRFX_PROJECT_` table would be missing when trying to get a single object.

------------------------------------------------------------

Is this behaviour expected? I'm wondering because if we take a look at OpenCrx demo server (for brevity let's focus on the `gender` code) so we'll first call the gender `CodeValueContainer`:
`https://demo.opencrx.org/opencrx-rest-CRX/org.opencrx.kernel.code1/provider/CRX/segment/Root/valueContainer/gender`

I won't paste the output here but note that this result is already a single `CodeValueContainer` object and it does show attributes such as `createdBy` and `modifiedBy`.

Now, for better demonstration, let's go one level deep and inspect:

* all `CodeValueEntry`s: `https://demo.opencrx.org/opencrx-rest-CRX/org.opencrx.kernel.code1/provider/CRX/segment/Root/valueContainer/gender/entry`

* a single `CodeValueEntry` (1 => male):
`https://demo.opencrx.org/opencrx-rest-CRX/org.opencrx.kernel.code1/provider/CRX/segment/Root/valueContainer/gender/entry/1`
Note that the entry `/1` here contains attributes such as `shortText` and `longText` which are saved in the `PRFX_CODEVALUEENTRY_` table of OpenCRX

----------------------------------------------------------------

I encountered this issue while trying to reproduce codevalue-based functionality and was unable to understand what's causing this behaviour in my openmdx-based app.

**Comment by javac9** [comment]
@wfro64 Do you have any clues what Workshop example might be missing, causing multivalued attribues not be fetched?

**Comment by DlRTY-HARRY** [comment]
openMDX uses [JDO Fetch Groups](https://db.apache.org/jdo/api31/apidocs/javax/jdo/FetchGroup.html) to determine which members of an object shall be retrieved. If a request does not specify a specific fetch group then the members belonging to the [DEFAULT Category](https://db.apache.org/jdo/api31/apidocs/javax/jdo/FetchPlan.html#DEFAULT) of the given result set are included in the reply.

It is up to a given deployment to define define which members belong to the DEFAULT category, which may therefore include no, some or all multivalued members. However, a native openMDX client will fetch any missing members lazily upon access if they have not been provided by the initial reply.

If a client wants to retrieve all members eagerly—including the ones not belonging to the DEFAULT category—then the [ALL Category](https://db.apache.org/jdo/api31/apidocs/javax/jdo/FetchPlan.html#ALL) shall be specified in the request.

**Comment by javac9** [comment]
Thank you Harry for your help!
